javatools/src/main/java/org/xvm/asm/ClassStructure.java

```diff
@@ -4255,6 +4255,28 @@ else if (cFormal > cActual)
                     m_listActual = listActual = new ArrayList<>(listActual); // clone
                     List<Map.Entry<StringConstant, TypeConstant>> entries = getTypeParamsAsList();
 
+                    if (id.equals(pool.clzClass()) && cActual > 0)
+                        {
+                        TypeConstant typePublic = listActual.get(0);
+                        if (typePublic.isSingleUnderlyingClass(false) && !typePublic.isFormalType())
+                            {
+                            // we know a bit more about the relationship between Class formal types;
+                            // prime them accordingly
+                            for (int i = cActual; i < cFormal; i++)
+                                {
+                                listActual.add(pool.ensureAccessTypeConstant(typePublic,
+                                    switch (i)
+                                        {
+                                        case 1  -> Access.PROTECTED;
+                                        case 2  -> Access.PRIVATE;
+                                        case 3  -> Access.STRUCT;
+                                        default -> throw new IllegalStateException();
+                                        }));
```

Comment:
Any exception thrown just gets to the top as a launcher exception, and has no info in it. I strongly suggest that the IllegalStateException has a message object, or that we use our own CompilerException .

I also STRONGLY suggest that that exception is not a RuntimeException, because it is very hard to debug this code if you don't know it and runtime exceptions are thrown earlier in the flow. It further forces the user to declare throws clauses in methods, so that it can be seen this method propagates upwards and needs to be handled already. 

This has been one of my main issues when learning the fundamentals of the compiler and runner. If something is a runtime exception, it should really be asynchronous, random and not expected IMHO

Reply:
I would generally agree, but in this case this is basically an assert for a logically impossible scenario. The value of cFormal is known to be 4, the value of cActual is known to be greater than 0 and less than 4. The default case is just required by the switch expression.